source/extensions/quic_listeners/quiche/platform/quic_mem_slice_impl.h

```diff
@@ -0,0 +1,87 @@
+#pragma once
+
+// NOLINT(namespace-envoy)
+
+// This file is part of the QUICHE platform implementation, and is not to be
+// consumed or referenced directly by other Envoy code. It serves purely as a
+// porting layer for QUICHE.
+
+#include <iostream>
+#include <memory>
+
+#include "common/buffer/buffer_impl.h"
+#include "common/common/assert.h"
+
+#include "quiche/quic/core/quic_buffer_allocator.h"
+
+namespace quic {
+
+// Implements the interface required by
+// https://quiche.googlesource.com/quiche/+/refs/heads/master/quic/platform/api/quic_mem_slice.h
+// Note the Quic plugin strategy relies on link-time binding rather than inheritance.
+class QuicMemSliceImpl {
+public:
+  // Constructs an empty QuicMemSliceImpl.
+  QuicMemSliceImpl() = default;
+
+  // Constructs a QuicMemSliceImpl by let |allocator| allocate a data buffer of
+  // |length|.
+  QuicMemSliceImpl(QuicBufferAllocator* allocator, size_t length) {
+    size_t total_length = sizeof(Envoy::Buffer::BufferFragmentImpl) + length;
+    char* mem = allocator->New(total_length);
+    auto fragment = new (mem + length) Envoy::Buffer::BufferFragmentImpl(
+        mem, length,
+        [allocator](const void* data, size_t, const Envoy::Buffer::BufferFragmentImpl*) {
+          allocator->Delete(const_cast<char*>(static_cast<const char*>(data)));
+        });
+    single_slice_buffer_.addBufferFragment(*fragment);
+  }
+
+  // Constructs a QuicMemSliceImpl from a Buffer::Instance with first |length| bytes in it.
+  // Data will be moved from |buffer| to this mem slice.
+  // Prerequisite: |buffer| has at least |length| bytes of data and not empty.
+  explicit QuicMemSliceImpl(Envoy::Buffer::Instance& buffer, size_t length) {
+#ifndef NDEBUG
```

Comment:
Instead of the #ifndef, how about DCHECK(FirstSliceLength(buffer), length), where FirstSliceLength is a function to get the length of the first slice?

Reply:
done